src/Strategy/Dart/PubDeps.hs

```diff
@@ -0,0 +1,215 @@
+
+  -- package may not have any dependencies
+  packageSubDeps <- optional (between (symbol "[") (symbol "]") (Set.fromList <$> sepBy parsePackageName (symbol " ")))
+
+  _ <- optional newline
+  pure $ PubDepPackage packageName (Just packageVersion) packageSubDeps isDirectDep
+
+depsCmdOutputParser :: Parser [PubDepPackage]
+depsCmdOutputParser = parseDeps <* eof
+  where
+    parseDeps = do
+      directDeps <- skipManyTill anySingle (symbol "dependencies:") *> many (parsePubDepPackage True)
+      devDeps <- optional $ chunk "dev dependencies:" *> newline *> many (parsePubDepPackage True)
+
+      -- Since, this strategy requires lockfile - dependency override information is redundant,
+      -- as lock file already produces resolved dependency source.
+      _ <- optional $ chunk "dependency overrides:" *> newline *> many (parsePubDepPackage False)
+      transitiveDeps <- optional $ chunk "transitive dependencies:" *> newline *> many (parsePubDepPackage False)
+
+      pure $
+        directDeps
+          ++ fromMaybe [] devDeps
+          ++ fromMaybe [] transitiveDeps
+
+isPackageSupported :: PubLockContent -> PackageName -> Bool
+isPackageSupported lockContent pkg = maybe False isSupported $ Map.lookup pkg (packages lockContent)
+
+buildGraph :: PubLockContent -> [PubDepPackage] -> Graphing Dependency
+buildGraph lockContent pkgs = Graphing.promoteToDirect isDirectDependency graphPackagesWithEdges
+  where
+    isDirectDependency :: Dependency -> Bool
+    isDirectDependency dep = dependencyName dep `elem` (dependencyName . pkgToDependency <$> reportableDirectPackages)
+
+    graphPackagesWithEdges :: Graphing Dependency
+    graphPackagesWithEdges = foldr (uncurry edge) graphPackagesWithoutEdges edges
+
+    edges :: [(Dependency, Dependency)]
+    edges = concatMap edgesOf pkgs
+
+    edgesOf :: PubDepPackage -> [(Dependency, Dependency)]
+    edgesOf pkg =
+      (pkgToDependency $ pubDepPackageName pkg,) . pkgToDependency
+        <$> filter isReportable (toList (fromMaybe Set.empty $ pubDepPackageDeps pkg))
+
+    pkgToDependency :: PackageName -> Dependency
+    pkgToDependency pkg = toDependency pkg $ metadataOf pkg
+
+    metadataOf :: PackageName -> PubLockPackageMetadata
+    metadataOf pkg =
+      Map.findWithDefault
+        ( PubLockPackageMetadata
+            { pubLockPackageIsDirect = any (\x -> pubDepPackageName x == pkg) pkgs
+            , pubLockPackageSource = PubLockPackageHostedSource Nothing Nothing
+            , pubLockPackageVersion = pubDepPackageVersion =<< find (\x -> pubDepPackageName x == pkg) pkgs
+            , pubLockPackageEnvironment = []
+            }
+        )
+        pkg
+        (packages lockContent)
+
+    graphPackagesWithoutEdges :: Graphing Dependency
+    graphPackagesWithoutEdges = foldr deep Graphing.empty depsWithoutEdges
+
+    depsWithoutEdges :: [Dependency]
+    depsWithoutEdges =
+      map pkgToDependency $
+        filter isReportable $
```

Comment:
It might make it clearer to centralize the `isReportable` filtering -- because right now we're doing it in several places

Specifically, something like:
```hs
-- build up a graph of package names, including edges/direct/etc
graphOfPackageNames :: Graphing PackageName
graphOfPackageNames = ...

-- use the `shrink` function to filter out unreportable packages
filteredGraphOfPackageNames :: Graphing PackageName
filteredGraphOfPackageNames = Graphing.shrink (isReportable . pubDepPackageName) graphOfPackageNames

-- map in `pkgToDependency`
finalGraph :: Graphing Dependency
finalGraph = Graphing.gmap pkgToDependency filteredGraphOfPackageNames
```

---

In general, it's best practice to defer transforming nodes in our graph to `Dependency` until the very last moment, doing incremental transformations on `Graphing WellBehavedIntermediateTypes`

Reply:
I've modified this whole building of graph here in recent commit: https://github.com/fossas/spectrometer/pull/313/commits/83f5771f03c59ee7f69a3e6331b1e91eed6f5406